src/auth0-session/cookie-store.ts

```diff
@@ -49,20 +38,23 @@ export default class CookieStore {
     this.chunkSize = MAX_COOKIE_SIZE - emptyCookie.length;
   }
 
-  private encrypt(payload: string, headers: { [key: string]: any }): string {
-    return JWE.encrypt(payload, this.currentKey as JWK.OctKey, {
-      alg,
-      enc,
-      ...headers
-    });
+  private async getKeys(): Promise<Uint8Array[]> {
+    if (!this.keys) {
+      const secret = this.config.secret;
+      const secrets = Array.isArray(secret) ? secret : [secret];
+      this.keys = await Promise.all(secrets.map(deriveKey));
+    }
+    return this.keys;
+  }
+
+  private async encrypt(payload: jose.JWTPayload, { iat, uat, exp }: Header): Promise<string> {
+    const [key] = await this.getKeys();
+    return await new jose.EncryptJWT({ ...payload }).setProtectedHeader({ alg, enc, uat, iat, exp }).encrypt(key);
   }
 
-  private decrypt(jwe: string): JWE.completeDecrypt {
-    return JWE.decrypt(jwe, this.keystore, {
-      complete: true,
-      contentEncryptionAlgorithms: [enc],
-      keyManagementAlgorithms: [alg]
-    });
+  private async decrypt(jwe: string): Promise<jose.CompactDecryptResult> {
+    const keys = await this.getKeys();
+    return pAny(keys.map((key) => jose.compactDecrypt(jwe, key))) as Promise<jose.CompactDecryptResult>;
```

Comment:
I believe it would be more efficient to iterate over the individual keys until a success occurs than to fire away all decrypt operations at once. This can be accompanied by a doc recommendation to put the most recent key at the front of the array of keys (when array is used).

Also, any reason not to use `jwtDecrypt` instead of `compactDecrypt`?

Reply:
👍 005c8e4